calm_adapter/calm_adapter/src/test/scala/uk/ac/wellcome/calm_adapter/CalmStoreTest.scala

```diff
@@ -10,39 +11,44 @@ import uk.ac.wellcome.storage.{StoreReadError, Version}
 class CalmStoreTest extends FunSpec with Matchers {
 
   type Key = Version[String, Int]
-  type Data = Map[String, String]
+
+  val retrievedAt = Instant.ofEpochSecond(123456)
 
   it("stores new CALM records") {
     val data = dataStore()
-    val record = CalmRecord("A", Map("key" -> "value"))
+    val record = CalmRecord("A", Map("key" -> "value"), retrievedAt)
     calmStore(data).putRecord(record) shouldBe Right(Some(Version("A", 0)))
-    data.entries shouldBe Map(
-      Version("A", 0) -> Map("key" -> "value")
-    )
+    data.entries shouldBe Map(Version("A", 0) -> record)
   }
 
-  it("stores already seen CALM records when the data has changed") {
-    val data = dataStore(Map(Version("A", 1) -> Map("key" -> "old")))
-    val record = CalmRecord("A", Map("key" -> "new"))
-    calmStore(data).putRecord(record) shouldBe Right(Some(Version("A", 2)))
+  it("stores already seen CALM records when newer data") {
+    val oldRecord = CalmRecord("A", Map("key" -> "old"), retrievedAt)
+    val newRecord = CalmRecord(
+      "A",
+      Map("key" -> "new"),
+      Instant.ofEpochSecond(retrievedAt.getEpochSecond + 2))
+    val data = dataStore(Version("A", 1) -> oldRecord)
+    calmStore(data).putRecord(newRecord) shouldBe Right(Some(Version("A", 2)))
     data.entries shouldBe Map(
-      Version("A", 1) -> Map("key" -> "old"),
-      Version("A", 2) -> Map("key" -> "new")
+      Version("A", 1) -> oldRecord,
+      Version("A", 2) -> newRecord
     )
   }
 
-  it("doesn't store already seen CALM records when unchanged data") {
-    val data = dataStore(Map(Version("A", 4) -> Map("key" -> "new")))
-    val record = CalmRecord("A", Map("key" -> "new"))
-    calmStore(data).putRecord(record) shouldBe Right(None)
-    data.entries shouldBe Map(
-      Version("A", 4) -> Map("key" -> "new"),
-    )
+  it("doesn't store already seen CALM records when older data") {
+    val oldRecord = CalmRecord("A", Map("key" -> "old"), retrievedAt)
+    val newRecord = CalmRecord(
+      "A",
+      Map("key" -> "new"),
+      Instant.ofEpochSecond(retrievedAt.getEpochSecond + 2))
+    val data = dataStore(Version("A", 4) -> newRecord)
+    calmStore(data).putRecord(oldRecord) shouldBe Right(None)
+    data.entries shouldBe Map(Version("A", 4) -> newRecord)
```

Comment:
There should also be a test here for the boundary case: what if the retrieval date of the stored and new record are exactly the same?

Reply:
@alexwlchan I'm not actually sure what the correct behaviour is in this case (where the retrieval date is exactly the same to the second, but the data differs between the two). Maybe the most robust thing to do would be to raise an error, so that the message ends up being retried and we can hopefully resolve the latest data? Seems like such an edge case, but maybe worth it.

Reply:
Hmm. I would go the other way – maybe raise an error if the retrieval date matches *and* the data is different? If the data is the same, then it's just SQS double-sending a message, and we can quietly drop it.

Reply:
I was suggesting exactly the same thing, sorry if my wording was unclear!

Reply:
🤦‍♀ 

Yes you did, I suck at reading comprehension.

Okay, do that!

Reply:
Done here: https://github.com/wellcometrust/catalogue/pull/407/commits/d4162e5b3e73deb632fbb0c10733d7d4d970edbe